### ISSUE TITLE
Introduce usage of float math functions.

### ISSUE THREAD

**Issue by carlo-bramini** [FluidSynth/fluidsynth#537]
Hello,
I wanted to ask what do you think about adding support for float math functions when `WITH_FLOAT` is defined. In this case, Fluidsynth may use something like this:

```
#ifdef WITH_FLOAT
#define FSIN(_x)    sinf(_x)
#define FCOS(_x)    cosf(_a)
#define FLOG(_x)    logf(_x)
...etc
#else
#define FSIN(_x)    sin(_x)
#define FCOS(_x)    cos(_a)
#define FLOG(_x)    log(_x)
...etc
#endif
```
These macros to functions could be used when the required type is `fluid_real_t` and they may be able to improve a bit the performance: in the current code there are several points where the traditional functions will not provide any additional precision because the parameters are promoted to `double` but the result is typically converted to `float`.
Of course, cmake is supposed to test if these functions exist before using them.
Opinions?

Sincerely.

**Comment by jjceresa** [comment]
Hello,
Assuming this will be supported, the corresponding CMake option should have a "performance optimization" meaning. I just want to say that the user shouldn't be confused between "precision" meaning given by actual `enable-floats` option and "optimization" option given by this new feature.

**Comment by derselbst** [comment]
> Opinions?

Makes sense to me, esp. for embedded system which may have their own optimized implementation of those C99 single precision functions. IIR filter, chorus and reverb come to my mind, anything else?

Seems like a good proposal to me. For consistency however we should call it `FLUID_SIN `rather than `FSIN`.

IMO it would be best to use `CMAKE_C_COMPILE_FEATURES` and search for `c_std_99`. Unfortunately this would require to bump `cmake_minimum_required` to 3.8. Too big step. Perhaps better use [CheckSymbolExists](https://cmake.org/cmake/help/latest/module/CheckSymbolExists.html#module:CheckSymbolExists) whether sinf, cosf and friends are available.

> the corresponding CMake option should have a "performance optimization" meaning

I don't see a need introduce more cmake build options here. If the user actively enables floats, he is aware of a potential loss in precision which on the other hand may cause some (severe) optimizations, e.g. avoiding the need to calculate trigonometric functions up to double precision while truncating the result to single precision.

**Comment by carlo-bramini** [comment]
> Seems like a good proposal to me. For consistency however we should call it FLUID_SIN rather than FSIN.

Ok.

> I don't see a need introduce more cmake build options here. If the user actively enables floats, he is aware of a potential loss in precision.

This is what I was thinking.
Actually, writing:
`float v=(float)sin(angle);`
or
`float v=sinf(angle);`
will produce the same result because `v` is always truncated to float precision, but the second line is a bit faster.

Like you said, it was my intention to test the presence of those functions automatically if `WITH_FLOAT` is activated.
If a function is found, good.
If it is not, it can be always emulated with the usual call and a cast to float like now.

**Comment by mawe42** [comment]
> Like you said, it was my intention to test the presence of those functions automatically if WITH_FLOAT is activated. If a function is found, good. If it is not, it can be always emulated with the usual call and a cast to float like now.

Possible performance gain with a minimum of added complexity: Sounds very good to me!

**Comment by schnitzeltony** [comment]
I like the idea. My preference would be to enable it by WITH_FLOAT but I would be fine with a new option either.

**Comment by derselbst** [comment]
Considering this implemented by #545. Thanks @carlo-bramini!
